### PR TITLE
Add support for 16550 compatible PCI serial cards

### PR DESCRIPTION
**Pull request by szafonimateusz-mi**
## Summary
-  drivers: prepare 16550 UART driver to support PCI
    
    prepare 16550 UART driver to support PCI:
    
    - **[breaking change]** change argument of `uart_ioctl()` from `struct file *filep` to `FAR struct u16550_s *priv`
      Also fix moxart_16550.c build related to this change
    
    - **[breaking change]** change argument of `uart_getreg()` and `uart_putreg()` from `uart_addrwidth_t base` to `FAR struct u16550_s *priv`
      Also fix arch/x86/src/qemu/qemu_serial.c and arch/x86_64/src/intel64/intel64_serial.c related to this change
    
    - **[breaking change]** change argument of `uart_dmachan()` from `uart_addrwidth_t base` to `FAR struct u16550_s *priv`
    
    - move `struct u16550_s` to public header
    
    - generalize UART_XXX_OFFSET so we can use it with any register increment
    
    - make `u16550_bind()`, `u16550_interrupt()`, `u16550_interrupt()` public
    
    - remove arch/or1k/src/common/or1k_uart.c and use common 16550 MIMO interfacve
    
-   drivers: add UART 16550 compatible PCI device support
    add support for UART 16550 compatible PCI device  
    
 -   arch/intel64: add support for PCI serial
 
 -  boards/x86_64/qemu-intel64: add configs with PCI serial console
    add configs with PCI serial console for qemu-intel64 so we can run NuttX on bare
    metal Intel HW with serial port card
    
- drivers: fix gcc14 errors for PCI
    
[DONE] Depends on https://github.com/apache/nuttx/pull/13391

## Impact
out of tree arch that use UART 16550 may need to update interfaces used according to these changes:
- change argument of uart_ioctl() from `struct file *filep` to `FAR struct u16550_s *priv`
      look at moxart_16550.c modifed in this PR
    
- change argument of uart_getreg() and uart_putreg from `uart_addrwidth_t base` to `FAR struct u16550_s *priv`
      look at arch/x86/src/qemu/qemu_serial.c and arch/x86_64/src/intel64/intel64_serial.c modifed in this PR
    
- change argument of uart_dmachan() from `uart_addrwidth_t base` to `FAR struct u16550_s *priv`
  `uart_dmachan()` not used in upstream.

## Testing
tested on intel HW with serial PCI card and tested on qemu with command:
```
qemu-system-x86_64 -m 2048M -cpu host -enable-kvm -kernel nuttx.elf -nographic -chardev pty,id=ch1 -device pci-serial,chardev=ch1 
```

### CODE REVIEW

**Comment by raiden00pl** [comment]
@ldube FYI since it seems you are using a device with 16550 UART

**Comment by szafonimateusz-mi** [comment]
tested on qemu and real HW, it's ready to review

File: drivers/pci/Kconfig [hunk 1]
```diff
@@ -12,6 +12,14 @@ menuconfig PCI
 
 if PCI
 
+config PCI_LATE_DRIVERS_REGISTER
+	bool
+	default n
+	---help---
+		This option disable PCI drivers registratin from drivers_initialize().
```
Comment (ldube) [inline drivers/pci/Kconfig:19]: This option disables PCI driver registration in drivers_initialize()

Comment (ldube) [inline drivers/pci/Kconfig:19]: This option disables PCI driver registration in drivers_initialize()

**Review by lupyuen — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**